### PR TITLE
Implement LOR-to-HO transfer operator [lor2ho-dev]

### PR DESCRIPTION
**Pull request by v-dobrev**
The LOR-to-HO transfer operator is returned by the new method `FiniteElementSpace::GetReverseTransferOperator()`.

TODO:
- [x] Fix Travis errors
- [x] Update `CHANGELOG` 
- [x] Add an example?

### CODE REVIEW

**Comment by tzanio** [comment]
This is really cool, thanks Veselin!

@pazner, you may be interested to take a look

**Comment by tzanio** [comment]
Added some TODO items.

File: mesh/ncmesh.cpp [hunk 1]
```diff
@@ -31,6 +32,118 @@ const DenseTensor &CoarseFineTransformations::GetPointMatrices(
...
+         if (children[i].one < other.children[i].one) { return true; }
+         if (children[i].one > other.children[i].one) { return false; }
+      }
+      return false; // everything is equal
+   }
+};
+
+}
+
+void CoarseFineTransformations::GetCoarseToFineMap(
```
Comment (tzanio) [inline mesh/ncmesh.cpp]: Nice! This could be useful also in other contexts. (cc @jakubcerveny)

File: fem/fespace.hpp [hunk 2]
```diff
@@ -473,6 +495,19 @@ class FiniteElementSpace
    void GetTransferOperator(const FiniteElementSpace &coarse_fes,
                             OperatorHandle &T) const;
 
+   /** @brief Construct and return an Operator that can be used to transfer
+       GridFunction data from @a this FE space, defined on a fine mesh, to @a
+       coarse_fes, defined on a coarse mesh. */
+   /** This method returns a left inverse of the Operator returned by
+       GetTransferOperator().
+
```
Comment (tzanio) [inline fem/fespace.hpp]: Should we clarify that we assume that the operator (R) returned by `GetTransferOperator()` is full column rank? In particular we assume that the LOR space has dimension that is at least as big as that of the HO space.

**Comment by tzanio** [comment]
I tested this with the simple example below and it works really well!

ho-lor

One can adjust the orders of the HO and LOR spaces as well as the refinement levels of the mesh and things looks pretty good as long as the LOR space is large enough...

```c++
//                             MFEM Example HO/LOR
//
// Compile with: make lor
//
// Sample runs:  lor -lref 4 -o 4 -lo 0 -p 1
//               lor -lref 5 -o 4 -lo 0 -p 1
//               lor -lref 3 -o 4 -lo 0 -p 1
//               lor -lref 5 -o 4 -lo 2 -p 1
//               lor -lref 5 -o 4 -lo 3 -p 2
//               lor -lref 5 -o 4 -lo 0 -p 3
//
// Description:  HO <-> LOR tester

#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int problem = 1; // problem type

int Wx = 0, Wy = 0; // window position
// int Ww = 350, Wh = 350; // window size
// int offx = Ww+10, offy = Wh+45; // window offsets
int Ww = 286, Wh = 286; // window size
int offx = Ww+2, offy = Wh+25; // window offsets

// Exact functions to project
double V_exact(const Vector &x);
double RHO_exact(const Vector &x);

// Helper functions
void visualize(VisItDataCollection &, string, bool, bool, int, int, int, int);
void compute_mass(FiniteElementSpace *, FiniteElementSpace *, double, double,
                  VisItDataCollection &, string, bool, bool);

int main(int argc, char *argv[])
{
   // Parse command-line options.
   const char *mesh_file = "../data/star-q2.mesh";
   int order = 4;
   int lref = order;
   int lorder = 0;
   bool visualization = true;
   bool testH1 = false;
   bool testL2 = true;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&problem, "-p", "--problem",
                  "Problem type (see the *_exact functions.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree) or -1 for"
                  " isoparametric space.");
   args.AddOption(&lref, "-lref", "--lor-ref-level",
                  "LOR refinement level (equal to order means identity LOR).");
   args.AddOption(&lorder, "-lo", "--lor-order",
                  "LOR refinement order (polynomial degree, zero by default).");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&testH1, "-h1", "--test-h1", "-no-h1", "--dont-test-h1",
                  "Test H1 HO->LOR transfers.");
   args.AddOption(&testL2, "-l2", "--test-l2", "-no-l2", "--dont-test-l2",
                  "Test L2 HO->LOR transfers.");
   args.Parse();
   if (!args.Good())
   {
      args.PrintUsage(cout);
      return 1;
   }
   args.PrintOptions(cout);

   // Read the mesh from the given mesh file.
   Mesh *mesh = new Mesh(mesh_file, 1, 1);
   int dim = mesh->Dimension();

   // Refine the mesh to increase the resolution.
   {
      int ref_levels = 0;
      for (int l = 0; l < ref_levels; l++)
      {
         mesh->UniformRefinement();
      }
   }

   VisItDataCollection HO_dc("HO", mesh);

   FiniteElementCollection *fecH1 = new H1_FECollection(order, dim);
   FiniteElementCollection *fecL2 = new L2_FECollection(order-1, dim);
   FiniteElementSpace *fespaceH1 = new FiniteElementSpace(mesh, fecH1);
   FiniteElementSpace *fespaceL2 = new FiniteElementSpace(mesh, fecL2);

   GridFunction v(fespaceH1);
   GridFunction rho(fespaceL2);

   HO_dc.RegisterField("velocity", &v);
   HO_dc.RegisterField("density", &rho);

   cout.precision(12);

   double massL2;
   switch (problem)
   {
      case 1: massL2 = 0.11498649047418217639; break;
      case 2: massL2 = 8.43665192918e-06; break;
      case 3: massL2 = 3.2453694754931463606; break;
      default: massL2 = 4.755276696226997224;
   }
   double massH1 = massL2;

   if (testH1) { cout << "H1 Exact  mass   = " << massH1 << endl; }
   if (testL2) { cout << "L2 Exact  mass   = " << massL2 << endl; }

   // HO projections
   FunctionCoefficient V(V_exact), RHO(RHO_exact);
   {
      v.ProjectCoefficient(V);
      rho.ProjectCoefficient(RHO);

      if (visualization)
      {
         visualize(HO_dc, "HO", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespaceH1, fespaceL2, massH1, massL2,
                   HO_dc, "HO    ", testH1, testL2);
   }

   int basis_lor = BasisType::GaussLobatto; // BasisType::ClosedUniform;
   Mesh *mesh_lor = new Mesh(mesh, lref, basis_lor);

   VisItDataCollection LOR_dc("LOR", mesh_lor);

   FiniteElementCollection *fec_lorH1 = new H1_FECollection(lorder+1, dim);
   FiniteElementCollection *fec_lorL2 = new L2_FECollection(lorder, dim);
   FiniteElementSpace *fespace_lorH1 = new FiniteElementSpace(mesh_lor, fec_lorH1);
   FiniteElementSpace *fespace_lorL2 = new FiniteElementSpace(mesh_lor, fec_lorL2);

   GridFunction v_lor(fespace_lorH1);
   GridFunction rho_lor(fespace_lorL2);

   LOR_dc.RegisterField("velocity", &v_lor);
   LOR_dc.RegisterField("density", &rho_lor);

   // HO->LOR restriction
   OperatorHandle R_H1, R_L2;
   fespace_lorH1->GetTransferOperator(*fespaceH1, R_H1);
   fespace_lorL2->GetTransferOperator(*fespaceL2, R_L2);
   {
      R_H1.Ptr()->Mult(v, v_lor);
      R_L2.Ptr()->Mult(rho, rho_lor);

#if 0
      // Volume error correction
      for (int iHO = 0; iHO < mesh->GetNE(); iHO++)
      {
         for (int j = 0; j < order; j++)
         {
            for (int i = 0; i < order; i++)
            {
               int iLOR = iHO*order*order + j*order + i;
               double Vlor = mesh_lor->GetElementVolume(iLOR);
               double VHO = mesh->GetElementVolume(iHO);
               rho_lor(iLOR) *= VHO/Vlor / (double)(order*order);
            }
         }
      }
#endif

      if (visualization)
      {
         visualize(LOR_dc, "R(HO)", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespace_lorH1, fespace_lorL2, massH1, massL2,
                   LOR_dc, "R(HO) ", testH1, testL2);
   }

   // LOR -> HO prolongation, starting with the R(HO) restriction
   OperatorHandle P_H1, P_L2;
   fespace_lorH1->GetReverseTransferOperator(*(new MassIntegrator),
                                             *fespaceH1, P_H1);
   fespace_lorL2->GetReverseTransferOperator(*(new MassIntegrator),
                                             *fespaceL2, P_L2);
   {
      P_H1.Ptr()->Mult(v_lor, v);
      P_L2.Ptr()->Mult(rho_lor, rho);

      if (visualization)
      {
         visualize(HO_dc, "PR(HO)", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespaceH1, fespaceL2, massH1, massL2,
                   HO_dc, "PR(HO)", testH1, testL2);
   }

#if 0
   // LOR -> HO transfer based on D^{-1} R^T, where D = diag (R^T R 1)
   {
      GridFunction h1_one_ho(fespaceH1), h1_one_lor(fespace_lorH1);
      h1_one_ho = 1.0;
      R_H1.Ptr()->Mult(h1_one_ho, h1_one_lor);
      R_H1.Ptr()->MultTranspose(h1_one_lor, h1_one_ho);
      R_H1.Ptr()->MultTranspose(v_lor, v);
      for (int i = 0; i < rho.Size(); i++) { v(i) /= h1_one_ho(i); }

      GridFunction l2_one_ho(fespaceL2), l2_one_lor(fespace_lorL2);
      l2_one_ho = 1.0;
      R_L2.Ptr()->Mult(l2_one_ho, l2_one_lor);
      R_L2.Ptr()->MultTranspose(l2_one_lor, l2_one_ho);
      R_L2.Ptr()->MultTranspose(rho_lor, rho);
      for (int i = 0; i < rho.Size(); i++) { rho(i) /= l2_one_ho(i); }

      if (visualization)
      {
         visualize(HO_dc, "RtR(HO)", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespaceH1, fespaceL2, massH1, massL2,
                   HO_dc, "RtR(HO)", testH1, testL2);
   }
#endif

   // Direct LOR projection (not computable, just for comparison)
   {
      v_lor.ProjectCoefficient(V);
      rho_lor.ProjectCoefficient(RHO);

      if (visualization)
      {
         visualize(LOR_dc, "LOR", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespace_lorH1, fespace_lorL2, massH1, massL2,
                   LOR_dc, "LOR   ", testH1, testL2);
   }

   // LOR -> HO prolongation, starting with the direct LOR projection
   {
      P_H1.Ptr()->Mult(v_lor, v);
      P_L2.Ptr()->Mult(rho_lor, rho);

      if (visualization)
      {
         visualize(HO_dc, "P(LOR)", testH1, testL2, Wx, Wy, Ww, Wh);
         Wx += offx;
      }
      compute_mass(fespaceH1, fespaceL2, massH1, massL2,
                   HO_dc, "P(LOR)", testH1, testL2);
   }

   // Free the used memory.
   delete fespaceH1;
   delete fespaceL2;
   delete fecH1;
   delete fecL2;
   delete mesh;

   delete fespace_lorH1;
   delete fespace_lorL2;
   delete fec_lorH1;
   delete fec_lorL2;
   delete mesh_lor;

   return 0;
}


double V_exact(const Vector &x)
{
   return RHO_exact(x);
}


double RHO_exact(const Vector &x)
{
   switch (problem)
   {
      case 1: // smooth field
         return x(1)+0.25*cos(2*M_PI*x.Norml2());
      case 2: // cubic function
         return x(1)*x(1)*x(1) + 2*x(0)*x(1) + x(0);
      case 3: // sharp gradient
      {
         return M_PI/2-atan(5*(2*x.Norml2()-1));
      }
      default:
         return 1.0;
   }
}


void visualize(VisItDataCollection &dc, string prefix, bool testH1, bool testL2,
               int x, int y, int w, int h)
{
   char vishost[] = "localhost";
   int  visport   = 19916;

   if (testH1)
   {
      socketstream sol_sockH1(vishost, visport);
      sol_sockH1.precision(8);
      sol_sockH1 << "solution\n" << *dc.GetMesh() << *dc.GetField("velocity")
                 << "window_geometry " << x << " " << y+offy << " " << w << " " << h
                 << "plot_caption 'H1 " << prefix << " Velocity'"
                 << "window_title 'H1 " << prefix << " Velocity'" << flush;
   }

   if (testL2)
   {
      socketstream sol_sockL2(vishost, visport);
      sol_sockL2.precision(8);
      sol_sockL2 << "solution\n" << *dc.GetMesh() << *dc.GetField("density")
                 << "window_geometry " << x << " " << y << " " << w << " " << h
                 << "plot_caption 'L2 " << prefix << " Density'"
                 << "window_title 'L2 " << prefix << " Density'" << flush;
   }
}


void compute_mass(FiniteElementSpace *H1, FiniteElementSpace *L2,
                  double massH1, double massL2,
                  VisItDataCollection &dc, string prefix,
                  bool testH1, bool testL2)
{
   ConstantCoefficient one(1.0);

   if (testH1)
   {
      BilinearForm *MH1 = new BilinearForm(H1);
      MH1->AddDomainIntegrator(new MassIntegrator(one));
      MH1->Assemble();

      GridFunction vone(H1);
      vone = 1.0;

      double newmass = MH1->InnerProduct(*dc.GetField("velocity"),vone);
      cout.precision(12);
      cout << "H1 " << prefix << " mass   = " << newmass;
      cout.precision(4);
      cout << " (" << fabs(newmass-massH1)*100/massH1 << "%)" << endl;

      delete MH1;
   }

   if (testL2)
   {
      BilinearForm *ML2 = new BilinearForm(L2);
      ML2->AddDomainIntegrator(new MassIntegrator(one));
      ML2->Assemble();

      GridFunction rhoone(L2);
      rhoone = 1.0;

      double newmass = ML2->InnerProduct(*dc.GetField("density"),rhoone);
      cout.precision(12);
      cout << "L2 " << prefix << " mass   = " << newmass;
      cout.precision(4);
      cout << " ("  << fabs(newmass-massL2)*100/massL2 << "%)" << endl;

      delete ML2;
   }
}
```

File: fem/fespace.hpp (hunk 2 above)
Comment (rieben1) [reply fem/fespace.hpp]: I agree it would be good to make this explicit, perhaps even issuing a warning (error?) if this not the case.

**Comment by rieben1** [comment]
Thanks @v-dobrev!

**Comment by pazner** [comment]
Running in debug mode gave an out-of-bounds error at the std::sort call in GetCoarseToFineMap. I pushed a fix that I think addresses this

**Comment by pazner** [comment]
I added another alternative that may be useful when working with L2 spaces. The restriction operator R from HO to LOR is given by L2 projection (defined by the L2Projection class). The reverse operator P (defined by the L2Prolongation class) is of the form suggested @v-dobrev. We have P = (R^T M_LOR R)^{-1} R^T M_LOR, which ensures that P is a left-inverse of R (so restricting and then prolongating gives the original high-order function). It can also be shown that this P is mass conservative.

**Comment by pazner** [comment]
A slight modification of @tzanio's example can be used to verify these properties. Sample output:
l2projection
```
/l2proj -lo 1
Options used:
   --mesh ../data/star.mesh
   --problem 1
   --order 4
   --lor-ref-level 4
   --lor-order 1
   --visualization
   --test-l2
L2 Exact     mass   = 0.114986490474
L2 HO        mass   = 0.114984920888 (0.001365%)
L2 R(HO)     mass   = 0.114984920888 (0.001365%)
L2 P(R(HO))  mass   = 0.114984920888 (0.001365%)
HO - P(R(HO))       = 1.11022302463e-15
L2 LOR       mass   = 0.114963186921 (0.02027%)
L2 P(LOR)    mass   = 0.114963186921 (0.02027%)
LOR - R(P(LOR))     = 0.006896
```
Here is the code:
```c++
#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int problem = 1; // problem type

int Wx = 0, Wy = 0; // window position
// int Ww = 350, Wh = 350; // window size
// int offx = Ww+10, offy = Wh+45; // window offsets
int Ww = 286, Wh = 286; // window size
int offx = Ww+2, offy = Wh+25; // window offsets

// Exact functions to project
double RHO_exact(const Vector &x);

// Helper functions
void visualize(VisItDataCollection &, string, int, int, int, int);
void compute_mass(FiniteElementSpace *, double, VisItDataCollection &, string);

int main(int argc, char *argv[])
{
   // Parse command-line options.
   const char *mesh_file = "../data/star.mesh";
   int order = 4;
   int lref = order;
   int lorder = 0;
   bool visualization = true;
   bool testL2 = true;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&problem, "-p", "--problem",
                  "Problem type (see the *_exact functions.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree) or -1 for"
                  " isoparametric space.");
   args.AddOption(&lref, "-lref", "--lor-ref-level",
                  "LOR refinement level (equal to order means identity LOR).");
   args.AddOption(&lorder, "-lo", "--lor-order",
                  "LOR refinement order (polynomial degree, zero by default).");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&testL2, "-l2", "--test-l2", "-no-l2", "--dont-test-l2",
                  "Test L2 HO->LOR transfers.");
   args.Parse();
   if (!args.Good())
   {
      args.PrintUsage(cout);
      return 1;
   }
   args.PrintOptions(cout);

   // Read the mesh from the given mesh file.
   Mesh mesh(mesh_file, 1, 1);
   int dim = mesh.Dimension();

   // Refine the mesh to increase the resolution.
   {
      int ref_levels = 0;
      for (int l = 0; l < ref_levels; l++)
      {
         mesh.UniformRefinement();
      }
   }

   VisItDataCollection HO_dc("HO", &mesh);

   L2_FECollection fecL2(order-1, dim);
   FiniteElementSpace fespaceL2(&mesh, &fecL2);

   GridFunction rho(&fespaceL2);

   HO_dc.RegisterField("density", &rho);

   cout.precision(12);

   double massL2;
   switch (problem)
   {
      case 1: massL2 = 0.11498649047418217639; break;
      case 2: massL2 = 8.43665192918e-06; break;
      case 3: massL2 = 3.2453694754931463606; break;
      default: massL2 = 4.755276696226997224;
   }

   if (testL2) { cout << "L2 Exact     mass   = " << massL2 << endl; }

   // HO projections
   FunctionCoefficient RHO(RHO_exact);
   rho.ProjectCoefficient(RHO);
   if (visualization)
   {
      visualize(HO_dc, "HO", Wx, Wy, Ww, Wh);
      Wx += offx;
   }
   compute_mass(&fespaceL2, massL2, HO_dc, "HO       ");

   // Create LOR space
   int basis_lor = BasisType::GaussLobatto; // BasisType::ClosedUniform;
   Mesh mesh_lor(&mesh, lref, basis_lor);

   VisItDataCollection LOR_dc("LOR", &mesh_lor);

   L2_FECollection fec_lorL2(lorder, dim);
   FiniteElementSpace fespace_lorL2(&mesh_lor, &fec_lorL2);

   GridFunction rho_lor(&fespace_lorL2);

   LOR_dc.RegisterField("density", &rho_lor);

   // HO->LOR restriction
   L2Projection R(fespaceL2, fespace_lorL2);
   R.Mult(rho, rho_lor);
   if (visualization)
   {
      visualize(LOR_dc, "R(HO)", Wx, Wy, Ww, Wh);
      Wx += offx;
   }
   compute_mass(&fespace_lorL2, massL2, LOR_dc, "R(HO)    ");

   // LOR-HO prolongation
   GridFunction rho_prev = rho;
   L2Prolongation P(R);
   P.Mult(rho_lor, rho);

   if (visualization)
   {
      visualize(HO_dc, "P(R(HO))", Wx, Wy, Ww, Wh);
      Wx += offx;
   }
   compute_mass(&fespaceL2, massL2, HO_dc, "P(R(HO)) ");

   rho_prev -= rho;
   cout.precision(12);
   cout << "HO - P(R(HO))       = " << rho_prev.Normlinf() << endl;

   Wx = 0;
   Wy += offy;

   // LOR projections
   rho_lor.ProjectCoefficient(RHO);
   GridFunction rho_lor_prev = rho_lor;

   if (visualization)
   {
      visualize(LOR_dc, "LOR", Wx, Wy, Ww, Wh);
      Wx += offx;
   }
   compute_mass(&fespace_lorL2, massL2, LOR_dc, "LOR      ");

   // Prolongate to HO space
   P.Mult(rho_lor, rho);

   if (visualization)
   {
      visualize(HO_dc, "P(LOR)", Wx, Wy, Ww, Wh);
      Wx += offx;
   }
   compute_mass(&fespaceL2, massL2, HO_dc, "P(LOR)   ");

   // Restrict back to LOR space. This won't give the original function
   // because the rho_lor doesn't necessarily live in the range of R.
   R.Mult(rho, rho_lor);
   rho_lor_prev -= rho_lor;
   cout << "LOR - R(P(LOR))     = " << rho_lor_prev.Normlinf() << endl;

   return 0;
}


double RHO_exact(const Vector &x)
{
   switch (problem)
   {
      case 1: // smooth field
         return x(1)+0.25*cos(2*M_PI*x.Norml2());
      case 2: // cubic function
         return x(1)*x(1)*x(1) + 2*x(0)*x(1) + x(0);
      case 3: // sharp gradient
      {
         return M_PI/2-atan(5*(2*x.Norml2()-1));
      }
      default:
         return 1.0;
   }
}


void visualize(VisItDataCollection &dc, string prefix,
               int x, int y, int w, int h)
{
   char vishost[] = "localhost";
   int  visport   = 19916;

   socketstream sol_sockL2(vishost, visport);
   sol_sockL2.precision(8);
   sol_sockL2 << "solution\n" << *dc.GetMesh() << *dc.GetField("density")
              << "window_geometry " << x << " " << y << " " << w << " " << h
              << "plot_caption 'L2 " << prefix << " Density'"
              << "window_title 'L2 " << prefix << " Density'" << flush;
}


void compute_mass(FiniteElementSpace *L2, double massL2,
                  VisItDataCollection &dc, string prefix)
{
   ConstantCoefficient one(1.0);
   BilinearForm *ML2 = new BilinearForm(L2);
   ML2->AddDomainIntegrator(new MassIntegrator(one));
   ML2->Assemble();

   GridFunction rhoone(L2);
   rhoone = 1.0;

   double newmass = ML2->InnerProduct(*dc.GetField("density"),rhoone);
   cout.precision(12);
   cout << "L2 " << prefix << " mass   = " << newmass;
   cout.precision(4);
   cout << " ("  << fabs(newmass-massL2)*100/massL2 << "%)" << endl;

   delete ML2;
}
```

**Comment by tzanio** [comment]
This looks great, thanks @v-dobrev and @pazner!

@vladotomov, can you take a look?

**Comment by vladotomov** [comment]
Impressive work, some very minor comments.

It's a little weird that similar things are done through different interfaces (getting operators from one of the FESpaces vs constructing them given the two spaces). I like the second approach more as it provides some separation from the FESpace class. Should be easy to move to a common interface, but it's also fine as it is.

Some of the comments claim that everything is conservative, but then turns out it's not always the case, this might be explained in the docs better.

**Pull request merged**